Layer refactor

SergioMartin86 opened the pull request:
+ Added RNN
+ Refactored layer code. It's more modular now
+ Refactored NN definition on examples so one simply define hidden layers
+ Refactored supervised learning exp definition in RL methods
+ Added time sequence support on NN
+ Improved checkpoint/resume logic
+ Removed unused RL algos
+ Changed RL algos to use timesequence (instead of individual states)
+ Added REFER as common tool for all algorithms to use
+ Milestone: GFPT solving ABF2D in 7 mins
+ Milestone: VRACER solving ABF3D
+ Added Retrace as common tool for all algorithms
+ Improved statistics

wadaniel commented on examples/learning/supervised/run-rnn.py:
```diff
@@ -0,0 +1,136 @@
+e["Solver"]["Learning Rate"] = 0.0001
+
+### Defining the shape of the neural network
+
```
P please check if this configuration makes sense to you, it looks good to me

webepasc replied:
I think we have to discuss this...

webepasc commented on examples/learning/reinforcement/cartpole/run-vracer-discrete.py:
```diff
@@ -34,45 +34,46 @@
 
-e["Solver"]["Critic"]["Neural Network"]["Layers"][3]["Type"] = "Layer/Dense"
-e["Solver"]["Critic"]["Neural Network"]["Layers"][3]["Activation Function"]["Type"] = "Elementwise/Linear"
+e["Solver"]["Neural Network"]["Hidden Layers"][3]["Type"] = "Layer/Activation"
```
I must say I find this confusing. Technically applying an activation function is not a layer...

webepasc commented on source/modules/neuralNetwork/layer/recurrent/README.rst:
```diff
@@ -0,0 +1,5 @@
+Recurrent
+*************
+
+Represent a hidden layer where a recurrent transformation is applied.    
```
We should add a proper documentation..

webepasc commented on source/modules/neuralNetwork/layer/recurrent/gru/README.rst:
```diff
@@ -0,0 +1,5 @@
+GRU
+*************
+
+Uses the Gated recurrent units (GRUs) architecture.
```
We should add a proper documentation..

webepasc reviewed (approved):
There is a lot of changes in there. I dislike some of the changes to layer. I think we should discuss it. Also, I am still not too happy about the removal of DDPG, ACER, and dueling DQN. I would rather prefer to push forward the automatic testing for the modules rather than removing them...

wadaniel commented:
I noticed that the naf cartpole experiment segfaults and vracer cartpole is not converging

SergioMartin86 commented:
> I noticed that the naf cartpole experiment segfaults and vracer cartpole is not converging

Fixed them both. The segfault was a wrong access to the circular buffer (the access was wrong, but there was a bug in cBuffer that made it not fail. After I fixed the bug in cBuffer, the wrong access segfaulted. They are both fixed now.)

As per convergence, there was a copy-paste mistake in both NAF and VRACER where the supervisor was using "MSE" instead of "Direct Gradients". Should be fixed now.
